programs/marginfi/src/state/panic_state.rs

```diff
@@ -0,0 +1,364 @@
+use crate::{MarginfiError, MarginfiResult};
+use anchor_lang::prelude::*;
+use marginfi_type_crate::types::PanicState;
+
+pub trait PanicStateImpl {
+    fn pause(&mut self, current_timestamp: i64) -> MarginfiResult;
+    fn unpause(&mut self);
+    fn unpause_if_expired(&mut self, current_timestamp: i64);
+}
+
+impl PanicStateImpl for PanicState {
+    fn pause(&mut self, current_timestamp: i64) -> MarginfiResult<()> {
+        // Clear existing pause if expired
+        self.unpause_if_expired(current_timestamp);
+
+        // Reset daily count if needed
+        if current_timestamp.saturating_sub(self.last_daily_reset_timestamp)
+            >= Self::DAILY_RESET_INTERVAL
+        {
+            self.daily_pause_count = 0;
+            self.last_daily_reset_timestamp = current_timestamp;
+        }
+
+        require!(
+            self.can_pause(current_timestamp),
+            MarginfiError::PauseLimitExceeded
+        );
+
+        // If already paused and not expired, treats this as an "extend" operation.
+        if self.is_paused_flag() && !self.is_expired(current_timestamp) {
+            self.pause_start_timestamp = self
```

Comment:
How does this work exactly? Let's say at `N+x` where `x<pause_duration` if we shift forward the `pause start timestamp` from `N` to `N+pause_duration`. I'm assuming that this doesn't deactivate the pause somehow from between `N+x -> N+pause_duration`?

Reply:
nit - Perhaps this is one reason using a `pause_start_timestamp` instead of an `end_timestamp` is less intuitive

Reply:
That's the goal, and for you to validate in your review! It does have the awkward side effect of start being in the future, which is indeed somewhat less intuitive.